index.d.ts

```diff
@@ -0,0 +1,36 @@
+interface Options {
+	/**
+	 * Number of concurrently pending promises returned by `mapper`.
+	 *
+	 * @default Infinity
+	 */
+	concurrency?: number;
+}
+
+type Mapper<T, U> = (input: T, index: number) => U | Promise<U>
+
+/**
+ * Returns a `Promise` that is fulfilled when all promises in `input` and ones returned from `mapper` are fulfilled, or rejects if any of the promises reject. The fulfilled value is an `Array` of the fulfilled values returned from `mapper` in `input` order.
+ *
+ * @param input - Iterated over concurrently in the `mapper` function.
+ * @param mapper - Function which is called for every item in `input`. Expected to return a `Promise` or value.
+ * @param options - Options-object.
+ *
+ * @example
+ *
+ * const sites = [
+ * 	'ava.li',
+ * 	'todomvc.com
+ * ];
+ *
+ * (async () => {
+ * 	const mapper = async site => {
+ * 		const {requestUrl} = await got.head(site);
+ * 		return requestUrl;
+ * 	};
+ *
+ * 	const result = await pMap(sites, mapper, {concurrency: 2});
+ * 	//=> ['http://ava.li/', 'http://todomvc.com/']
+ * })();
+ */
+export default function <T, U>(input: Iterable<T>, mapper: Mapper<T, U>, options?: Options): Promise<U[]>;
```

Comment:
What do you think about using more descriptive names for `T` and `U`? I know it's the convention, but it hurts readability.

Here it could maybe be:

- `T` => `Element`
- `U` => `NewElement`

Reply:
Good point! I don't mind for 1 or 2 generics, but if you go to a list of more then it's not really clear anymore of what is what. Changed it.